src/Entity/Payment.php

```diff
@@ -281,7 +281,10 @@ public function setPaymentMethod(PluginPaymentMethodInterface $payment_method) {
     /** @var \Drupal\plugin\Plugin\Field\FieldType\PluginCollectionItemInterface $field_item */
     $this->get('payment_method')->applyDefaultValue();
     $this->get('payment_method')->appendItem($payment_method);
-
+    // The payment method might have been unserialized with an old payment
+    // object, trying to save that as new will result in exceptions. Set the
+    // current object again.
+    $this->getPaymentMethod()->setPayment($this);
```

Comment:
Call $payment_method->setPayment($this) instead, possibly before appending it, shouldn't make a difference but makes a bit more sense when reading.